### ISSUE TITLE
LPython bogusly changes attribute value

### ISSUE THREAD

**Issue by rebcabin** [lcompilers/lpython#1981]
repro: https://github.com/rebcabin/lpython/tree/brian-lasr/lasr/LP-pycharm/Issue1981

This is related to Issue #1978; i'm digging in to find out what's wrong. I reduced the test case to the bare minimum that illustrates the problem. 

The problem is that when I initialize a StringIO object from inside a function, an attribute value gets bogusly changed to a random number.

Consider the following code in `Issue1981/lasr_lexer.py`

```python
from lpython import (dataclass, InOut, i32,)
from lasr_stringio import (StringIO, __post_init__,)

NEWLINE      : str       = "\n"
WHITESPACE   : list[str] = [" ", ",", NEWLINE, "\r", "\t"]
EOF_SIGNAL   : str       = ''

@dataclass
class LasrLexer:
    fd   : StringIO = StringIO('', 0, 0) # prefer Optional[StringIO]
    line : i32 = 1 # prefer Line
    col  : i32 = 1 # prefer Col
    type : str = '' # prefer LTType
    val  : str = '' # prefer LTVal

def lexer_test():
    print('LEXER TEST')
    lexer : LasrLexer = LasrLexer(StringIO('  foo  bar  '))
    __post_init__(lexer.fd)

if __name__ == '__main__':
    lexer_test()
```

When I run it with LPython, I get

```bash
/Users/brian/CLionProjects/lpython/src/bin/python /Users/brian/CLionProjects/lpython/lasr/LP-pycharm/Issue1981/lasr_lexer.py 
LEXER TEST
StringIO __post_init__ preconditions:
    _buf =   foo  bar  .
    _len = 1.
    _0cursor = 1830662080.  ################ ATTENTION #############
StringIO __post_init__ postconditions:
    _buf =   foo  bar  .
    _len = 12.
    _0cursor = 1830662080.
```

When I change the code to the following, just moving the initialization inside main and out of the function `lexer_test`, I get the correct value for `_0cursor`:

```python
from lpython import (dataclass, InOut, i32,)
from lasr_stringio import (StringIO, __post_init__,)


NEWLINE      : str       = "\n"
WHITESPACE   : list[str] = [" ", ",", NEWLINE, "\r", "\t"]
EOF_SIGNAL   : str       = ''


@dataclass
class LasrLexer:
    fd   : StringIO = StringIO('', 0, 0) # prefer Optional[StringIO]
    line : i32 = 1 # prefer Line
    col  : i32 = 1 # prefer Col
    type : str = '' # prefer LTType
    val  : str = '' # prefer LTVal


if __name__ == '__main__':
    print('LEXER TEST')
    lexer : LasrLexer = LasrLexer(StringIO('  foo  bar  '))
    __post_init__(lexer.fd)
```

produces: 

```bash
/Users/brian/CLionProjects/lpython/src/bin/python /Users/brian/CLionProjects/lpython/lasr/LP-pycharm/Issue1981/lasr_lexer.py 
LEXER TEST
StringIO __post_init__ preconditions:
    _buf =   foo  bar  .
    _len = 0.
    _0cursor = 0.  ################ ATTENTION ###############
StringIO __post_init__ postconditions:
    _buf =   foo  bar  .
    _len = 12.
    _0cursor = 0.
```

**Comment by rebcabin** [comment]
This could be related to a recent fix that allowed constructors to be called with partial argument lists. Investigating ...

**Comment by Smit-create** [comment]
```py

@dataclass
class StringIO:
    _buf     : str
    _0cursor : i32 = 0
    _len     : i32 = 0
```

Using this will fix.

**Comment by rebcabin** [comment]
ok, that's fair, but shouldn't `i32(0)` be allowed as an initial value?

**Comment by rebcabin** [comment]
Also, this fixes it too:

```python
    lexer : LasrLexer = LasrLexer(StringIO('  foo  bar  ', 0, 0))
```

instead of 

```python
    lexer : LasrLexer = LasrLexer(StringIO('  foo  bar  '))
```

probably for the same reason as your fix works.

**Comment by rebcabin** [comment]
This also fixes it:

```python
    lexer : LasrLexer = LasrLexer(StringIO('  foo  bar  ', i32(0), i32(0)))
```

So I think this bug still stands. The following _ought_ to work:

```python
@dataclass
class StringIO:
    _buf     : str
    _0cursor : i32 = i32(0)
    _len     : i32 = i32(0)
```

**Comment by rebcabin** [comment]
and the following _ought_ to work, too:

```python
    lexer : LasrLexer = LasrLexer(StringIO('  foo  bar  '))
```

**Comment by rebcabin** [comment]
The fact that the original code fails with a large random number that looks like a 64-bit address is evidence of a pointer calculation going wrong in the C++ code :)

**Comment by rebcabin** [comment]
I'm leaving the repro case untouched for investigation purposes.

**Comment by Thirumalai-Shaktivel** [comment]
Yup, it can also mean that the value `_0cursor : i32 = i32(0)` is not getting initialised!

**Comment by rebcabin** [comment]
you guys are great! this is fun!

**Comment by Thirumalai-Shaktivel** [comment]
I see, anything apart from constants is skipped in the StructType definition:
```py
@dataclass
class StringIO:
    _buf     : str
    _0cursor : i32 = 0
    _len     : i32 = i32(0)
```
```clojure
(StructType
    (SymbolTable
        4
        {
            _0cursor:
                (Variable
                    4
                    _0cursor
                    []
                    Local
                    ()
                    (IntegerConstant 0 (Integer 4))
                    Default
                    (Integer 4)
                    ()
                    Source
                    Public
                    Required
                    .false.
                ),
            _buf:
                (Variable
                    4
                    _buf
                    []
                    Local
                    ()
                    ()
                    Default
                    (Character 1 -2 ())
                    ()
                    Source
                    Public
                    Required
                    .false.
                ),
            _len:
                (Variable
                    4
                    _len
                    []
                    Local
                    ()
                    ()
                    Default
                    (Integer 4)
                    ()
                    Source
                    Public
                    Required
                    .false.
                )
        })
    StringIO
    []
    [_buf
    _0cursor
    _len]
    Source
    Public
    .false.
    .false.
    ()
    ()
)
```

**Comment by Thirumalai-Shaktivel** [comment]
https://github.com/lcompilers/lpython/blob/bcd882a6cc8e8b0dbdbda931b2c4d75b6c3aa256/src/lpython/semantics/python_ast_to_asr.cpp#L1332

This becomes nullptr for i32(0)

**Comment by Smit-create** [comment]
Struct members for initialization at present need compile-time constant values. And that so that's the reason for failure.

**Comment by Smit-create** [comment]
> ok, that's fair, but shouldn't i32(0) be allowed as an initial value?

It should, I'll look into that for sure.

**Comment by Smit-create** [comment]
Okay, the bug here is that `i32(0)` is a Call AST node that is skipped while visiting the Symbol table, and so, the value for it is kept as `nullptr`.

**Comment by Smit-create** [comment]
One fix is to make the Classdef node revisit again in body-visitor to capture this.

**Comment by certik** [comment]
I think that expressions have to be visited in the symbol table visitor, just like we do in LFortran. Because as part of types as well as initializations you need to visit them.

**Comment by Smit-create** [comment]
Fixed in #2056
